In [221]:
import os
os.chdir('/Users/tinapraveen/Desktop')
os.getcwd()

'/Users/tinapraveen/Desktop'

# Read clean data set

In [222]:
%matplotlib notebook
import numpy as np
import pandas as pd
clean_df=pd.read_csv('communities-crime-clean.csv')
clean_df.head()


state      communityname  fold  population  householdsize  racepctblack  \
0      1      Alabastercity     7        0.01           0.61          0.21   
1      1  AlexanderCitycity    10        0.01           0.41          0.55   
2      1       Annistoncity     3        0.03           0.34          0.86   
3      1         Athenscity     8        0.01           0.38          0.35   
4      1         Auburncity     1        0.04           0.37          0.32   

   racePctWhite  racePctAsian  racePctHisp  agePct12t21         ...           \
0          0.83          0.02         0.01         0.41         ...            
1          0.57          0.01         0.00         0.47         ...            
2          0.30          0.04         0.01         0.41         ...            
3          0.71          0.04         0.01         0.39         ...            
4          0.70          0.21         0.02         1.00         ...            

   PctForeignBorn  PctBornSameState  PctSameHouse85  PctSameCity85  \
0            0.03              0.70            0.40           0.34   
1            0.00              0.93            0.66           0.82   
2            0.04              0.77            0.59           0.70   
3            0.03              0.78            0.56           0.67   
4            0.12              0.49            0.12           0.00   

   PctSameState85  LandArea  PopDens  PctUsePubTrans  LemasPctOfficDrugUn  \
0            0.57      0.05     0.06            0.01                  0.0   
1            0.84      0.11     0.03            0.01                  0.0   
2            0.64      0.06     0.11            0.04                  0.0   
3            0.71      0.09     0.05            0.00                  0.0   
4            0.15      0.09     0.09            0.01                  0.0   

   ViolentCrimesPerPop  
0                 0.06  
1                 0.14  
2                 1.00  
3                 0.23  
4                 0.15  

[5 rows x 104 columns]

# Question 1

1 a.	Create a new field “highCrime” which is true if the crime rate per capita (ViolentCrimesPerPop) is greater than 0.1, and false otherwise. What are the percentage of positive and negative instances in the dataset?

In [223]:
clean_df['highCrime'] =( clean_df['ViolentCrimesPerPop']>0.1).astype(int)
clean_df.head()
from __future__ import division
Percentage_Postiveinst = (sum(clean_df['highCrime'])/len(clean_df)) * 100
print("PERCENTAGE OF POSITIVE INSTANCE : ", Percentage_Postiveinst)
Percentage_Negativeinst = 100 - Percentage_Postiveinst
print ("PERCENTAGE OF NEGATIVE INSTANCE : " , Percentage_Negativeinst)

PERCENTAGE OF POSITIVE INSTANCE :  62.7195183141
PERCENTAGE OF NEGATIVE INSTANCE :  37.2804816859


1.b .DECISION TREE CLASSIFIER

In [263]:
np.random.seed(0)
from sklearn import tree
X = clean_df.drop(['communityname','ViolentCrimesPerPop','highCrime','state','fold'],axis=1)
Y = clean_df['highCrime']
dt_classifier = tree.DecisionTreeClassifier()
dt_classifier = dt_classifier.fit(X, Y)

In [266]:
#clean_df.head()
X.head()

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0        0.01           0.61          0.21          0.83          0.02   
1        0.01           0.41          0.55          0.57          0.01   
2        0.03           0.34          0.86          0.30          0.04   
3        0.01           0.38          0.35          0.71          0.04   
4        0.04           0.37          0.32          0.70          0.21   

   racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  \
0         0.01         0.41         0.49         0.26        0.21   
1         0.00         0.47         0.45         0.31        0.57   
2         0.01         0.41         0.42         0.27        0.59   
3         0.01         0.39         0.46         0.31        0.49   
4         0.02         1.00         1.00         1.00        0.14   

          ...           NumStreet  PctForeignBorn  PctBornSameState  \
0         ...                0.00            0.03              0.70   
1         ...                0.00            0.00              0.93   
2         ...                0.02            0.04              0.77   
3         ...                0.00            0.03              0.78   
4         ...                0.00            0.12              0.49   

   PctSameHouse85  PctSameCity85  PctSameState85  LandArea  PopDens  \
0            0.40           0.34            0.57      0.05     0.06   
1            0.66           0.82            0.84      0.11     0.03   
2            0.59           0.70            0.64      0.06     0.11   
3            0.56           0.67            0.71      0.09     0.05   
4            0.12           0.00            0.15      0.09     0.09   

   PctUsePubTrans  LemasPctOfficDrugUn  
0            0.01                  0.0  
1            0.01                  0.0  
2            0.04                  0.0  
3            0.00                  0.0  
4            0.01                  0.0  

[5 rows x 100 columns]

In [267]:
dt_predicted = dt_classifier.predict(X)
pd.crosstab(clean_df['highCrime'], dt_predicted, rownames=['highCrime'], colnames=['DecisionTree_predictions'])

DecisionTree_predictions    0     1
highCrime                          
0                         743     0
1                           0  1250

1.b.i) DECISION TREE - ACCURACY , PRECISION AND RECALL

In [272]:
from sklearn.metrics import precision_score,recall_score,accuracy_score
dt_accuracy= accuracy_score(Y, dt_predicted)
dt_precision = precision_score( Y, dt_predicted)
dt_recall = recall_score( Y, dt_predicted)

print( "Accuracy = " , dt_accuracy ," Precision = ", dt_precision, ", Recall =" ,dt_recall)

Accuracy =  1.0  Precision =  1.0 , Recall = 1.0


1.b.ii)	What are the main features used for classification?

In [269]:
importances = dt_classifier.feature_importances_
importances = importances[importances!=0]
featurenames = np.argsort(importances)[::-1]

In [270]:
for k in range(5):
    print("%d.  %s %f" % (k + 1, X.columns[featurenames[k]], importances[featurenames[k]]))

1.  PctEmplProfServ 0.360083
2.  racePctWhite 0.088817
3.  racePctHisp 0.048585
4.  indianPerCap 0.023569
5.  NumUnderPov 0.019933


1.b.ii)Can you explain why they make sense (or not)

Top 5 features are extracted and below given are the descriptions:

PctOccupMgmtProf: percentage of people 16 and over who are employed in management or professional occupations
racePctWhite: percentage of population that is caucasian
racepctHiso: percentage of population that is  Hispanic
NumUnderPov number of people under the poverty level
blackPerCap: per capita income for african americans 

Let us analyze the above features:


In [274]:
Lowcrimerate_PctEmplProfServ= clean_df.loc[clean_df['highCrime'] == 0, 'PctEmplProfServ'].mean()
#Lowcrimerate_PctEmplProfServ = clean_df.loc[clean_df['highCrime'] == 0, 'Lowcrimerate_PctEmplProfServ'].sum()
HighCrimerate_PctEmplProfServ = clean_df.loc[clean_df['highCrime'] == 1, 'PctEmplProfServ'].mean()
print(" Low Crime rate among those who are employed in professional services:", Lowcrimerate_PctEmplProfServ)
print("  HighCrime rate among those who are employed in in professional services:", HighCrimerate_PctEmplProfServ)

 Low Crime rate among those who are employed in professional services: 0.47113055181695823
  HighCrime rate among those who are employed in in professional services: 0.4223759999999999


In [275]:
Nocrimerate_racePctWhite= clean_df.loc[clean_df['highCrime'] == 0, 'racePctWhite'].mean()
Crimerate_racePctWhite = clean_df.loc[clean_df['highCrime'] == 1, 'racePctWhite'].mean()
print(" Low Crime rate among percentage of population that is caucasian:", Nocrimerate_racePctWhite)
print(" HighCrime rate among percentage of population that is caucasian:", Crimerate_racePctWhite)


Nocrimerate_racepctHisp= clean_df.loc[clean_df['highCrime'] == 0, 'racePctHisp'].mean()
crimerate_racepctHisp = clean_df.loc[clean_df['highCrime'] == 1, 'racePctHisp'].mean()
print(" Low Crime rate among percentage of population that is  Hispanic:", Nocrimerate_racepctHisp)
print(" High Crime rate among percentage of population that is  Hispanic :", crimerate_racepctHisp)


Nocrimerate_NumUnderPov= clean_df.loc[clean_df['highCrime'] == 0, 'NumUnderPov'].mean()
crimerate_NumUnderPov = clean_df.loc[clean_df['highCrime'] == 1, 'NumUnderPov'].mean()
print(" Low Crime rate among  number of people under the poverty level:", Nocrimerate_NumUnderPov)
print(" High Crime rate among number of people under the poverty level :", crimerate_NumUnderPov)



Nocrimerate_indianPerCap= clean_df.loc[clean_df['highCrime'] == 0, 'indianPerCap'].mean()
crimerate_indianPerCap = clean_df.loc[clean_df['highCrime'] == 1, 'indianPerCap'].mean()
print(" LowCrime rate based on  per capita income for indians :", Nocrimerate_indianPerCap)
print(" HighCrime rate based on  per capita income for indians  :", crimerate_indianPerCap)





 Low Crime rate among percentage of population that is caucasian: 0.9134589502018869
 HighCrime rate among percentage of population that is caucasian: 0.6591920000000008
 Low Crime rate among percentage of population that is  Hispanic: 0.0502691790040378
 High Crime rate among percentage of population that is  Hispanic : 0.19985599999999976
 Low Crime rate among  number of people under the poverty level: 0.012893674293405011
 High Crime rate among number of people under the poverty level : 0.08084000000000016
 LowCrime rate based on  per capita income for indians : 0.21449528936742934
 HighCrime rate based on  per capita income for indians  : 0.19707199999999983


From the above numbers it can be infered that 
crime rate is lesser in areas where there are  people who are employed in  professional services. This can be justified as this groupo will be educated and hence chances are less to be engaged in criminal activities than the illiterates.

It can be seen that among caucasian the crime rate is less and crime rate is more in case of hispanic populated areas. Crime rate is more where there are people under the poverty line and hence the important features make sense.

Based on per capita income of indians NO crime rate is more.

1.c. i)CROSS VALIDATION

In [231]:
from sklearn.model_selection import cross_val_score
dt_cv_accuracy = cross_val_score(dt_classifier, X, Y,cv=10,scoring="accuracy")
dt_cv_precision = cross_val_score(dt_classifier, X, Y,cv=10,scoring="precision")
dt_cv_recall = cross_val_score(dt_classifier, X, Y,cv=10,scoring="recall")

In [232]:
print ("CrossValidated_Accuracy = ", dt_cv_accuracy.mean())
print ("CrossValidated_Precision =", dt_cv_precision.mean())
print ("CrossValidated_Recall = ", dt_cv_recall.mean())

CrossValidated_Accuracy =  0.732022613065
CrossValidated_Precision = 0.78978369709
CrossValidated_Recall =  0.7776


1.c. ii) Why are they different from the results in the previous test?

First we fitted the decision tree classifier on the data which resulted in overfitting and hence accuracy is 1. Now , we have done 10 fold cross validation and which divides the data into 10 folds and leave one out as test fold. This will prevent overfitting of the data and hence we have different values after cross validation.

2.Linear Classification 
2.a.Gaussian NB

In [233]:
from sklearn.naive_bayes import GaussianNB
nb_classifier =GaussianNB()
nb_classifier.fit(X,Y)
nb_predicted=nb_classifier.predict(X)
precision = precision_score( true_values, nb_predicted)
recall = recall_score( true_values, nb_predicted)
accuracy= accuracy_score( true_values, nb_predicted)
print( " Accuracy = ",accuracy,"Precision = ",precision,", Recall = ",recall)
pd.crosstab(clean_df['highCrime'], nb_predicted, rownames=['highCrime'], colnames=['NB_Predictions'])

 Accuracy =  0.778223783241 Precision =  0.931623931624 , Recall =  0.6976


NB_Predictions    0    1
highCrime               
0               679   64
1               378  872

2.a.i.	What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [234]:
nb_cv_accuracy = cross_val_score(nb_classifier, X, Y,cv=10,scoring="accuracy")
nb_cv_precision = cross_val_score(nb_classifier, X, Y,cv=10,scoring="precision")
nb_cv_recall = cross_val_score(nb_classifier, X, Y,cv=10,scoring="recall")
print("CrossValidated Accuracy_NB :", nb_cv_accuracy.mean())
print("CrossValidated Precision_NB: " ,nb_cv_precision.mean())
print("CrossValidated Recall_NB: ",  nb_cv_recall.mean())

CrossValidated Accuracy_NB : 0.761608040201
CrossValidated Precision_NB:  0.911799814828
CrossValidated Recall_NB:  0.692


2.a.ii.	What are the 10 most predictive features?

In [236]:
pred_features = pd.DataFrame({"Feature": [], "Score": []})
for k in X.columns:
    feature_mean = clean_df.groupby('highCrime')[k].mean()
    feature_var = clean_df.groupby('highCrime')[k].var()
    feature_sd = sum(np.sqrt(feature_var))
    pred_features = pred_features.append({"Feature": k, "Score": abs(feature_mean[0] - feature_mean[1])/feature_sd}, ignore_index=True)

iterations = np.argsort(feature_score['Score'])[::-1]
feature_score.iloc[iterations[0:10], ]

Feature     Score
46       PctKids2Par  0.809336
45        PctFam2Par  0.745162
5       racePctWhite  0.734884
52          PctIlleg  0.708929
42      FemalePctDiv  0.693604
43       TotalPctDiv  0.674282
47  PctYoungKids2Par  0.664671
17        pctWInvInc  0.660720
48       PctTeen2Par  0.642621
40    MalePctDivorce  0.616534

The larger this different, the more predictive the feature. Why do these make sense ?

Higher the normalized mean difference between the positive and negative class, higher the predictive power. Classification accuracy can be improved by defining threshold to classify the instances.

2.a.iii) How do these results compare with your results from decision trees?

the features which are scored by Naive Bayes are different from the important features from decision tree classifer.

Cross Validated  accuracy and Precision  of Guassian NB is more than that of Decision Tree classifier.But recall value of Gaussian NB is lesser than Decision Tree classifier. This means that NB underperforms Decision trees in case of classifying negative classes whereas decision trees have been consistent in classifying positive and negative instances.

2.b.LINEAR SVC

In [237]:
from sklearn.svm import LinearSVC
linearclassifier_SVC=LinearSVC()
linearclassifier_SVC.fit(X,Y)
svc_predicted=linearclassifier_SVC.predict(X)
precision = precision_score( true_values, svc_predicted)
recall = recall_score( true_values, svc_predicted)
accuracy= accuracy_score( true_values, svc_predicted)
print("Accuracy = ",accuracy,"Precision = ",precision, ", Recall = ",recall)
pd.crosstab(clean_df['highCrime'], svc_predicted, rownames=['highCrime'], colnames=['SVC_Predictions'])

Accuracy =  0.850476668339 Precision =  0.883870967742 , Recall =  0.8768


SVC_Predictions    0     1
highCrime                 
0                599   144
1                154  1096

2.b.i. What is the 10-fold cross-validation accuracy, precision, and recall for this method?

In [238]:
svc_cv_accuracy = cross_val_score(linearclassifier_SVC, X, Y,cv=10,scoring="accuracy")
svc_cv_precision = cross_val_score(linearclassifier_SVC, X, Y,cv=10,scoring="precision")
svc_cv_recall = cross_val_score(linearclassifier_SVC, X, Y,cv=10,scoring="recall")
print("CrossValidated Accuracy_SVC :", svc_cv_accuracy.mean())
print("CrossValidated Precision_SVC : " ,svc_cv_precision.mean())
print("CrossValidated Recall_SVC : ",  svc_cv_recall.mean())


CrossValidated Accuracy_SVC : 0.796233668342
CrossValidated Precision_SVC :  0.845404856531
CrossValidated Recall_SVC :  0.8344


2.b.ii.	What are the 10 most predictive features? This can be measured by the absolute feature weights in the model. 

In [239]:
feature_scores_SVC=abs(linearclassifier_SVC.coef_)[0]
iterations_SVC = np.argsort(feature_scores_SVC)[::-1]
for f in range(10):
    print("%d.  %s %f" % (f + 1, X.columns[iterations_SVC[f]], feature_scores_SVC[indices3[f]]))

1.  pctWInvInc 0.257815
2.  PersPerOccupHous 0.770301
3.  racePctWhite 0.379628
4.  PctKids2Par 0.273523
5.  RentHighQ 0.333322
6.  MalePctDivorce 0.759686
7.  NumUnderPov 0.200336
8.  NumStreet 0.019634
9.  PctOccupMgmtProf 0.404156
10.  population 0.009678


Why do these make sense (or not)?
 Let us inspect each of this features and take the mean of crime rate and No crime rate under these features. 

In [276]:
Nocrime_pctWInvInc = clean_df.loc[clean_df['highCrime'] == 0, 'pctWInvInc'].mean()
Crime_pctWInvInc = clean_df.loc[clean_df['highCrime'] == 1, 'pctWInvInc'].mean()

print ("Low Crime rate among Percentage of households with investment / rent income in 1989 ", Nocrime_pctWInvInc)
print ("HighCrime rate among Percentage of households with investment / rent income in 1989 ", Crime_pctWInvInc)


Nocrime_racePctWhite = clean_df.loc[clean_df['highCrime'] == 0, 'racePctWhite'].mean()
Crime_racePctWhite = clean_df.loc[clean_df['highCrime'] == 1, 'racePctWhite'].mean()

print ("Low crime rate in Percentage of population that is caucasian, ", Nocrime_racePctWhite)
print ("highCrime rate Percentage of population that is caucasian ", Crime_racePctWhite)



Nocrime_PctKids2Par = clean_df.loc[clean_df['highCrime'] == 0, 'PctKids2Par'].mean()
Crime_PctKids2Par = clean_df.loc[clean_df['highCrime'] == 1, 'PctKids2Par'].mean()


print ("Low crime rate in Percentage of kids in family housing with two parents: ", Nocrime_PctKids2Par)
print ("Highcrime rate in Percentage of kids in family housing with two parents: ", Crime_PctKids2Par)


Low Crime rate among Percentage of households with investment / rent income in 1989  0.6204845222072677
HighCrime rate among Percentage of households with investment / rent income in 1989  0.42165600000000003
Low crime rate in Percentage of population that is caucasian,  0.9134589502018869
highCrime rate Percentage of population that is caucasian  0.6591920000000008
Low crime rate in Percentage of kids in family housing with two parents:  0.7787617765814268
Highcrime rate in Percentage of kids in family housing with two parents:  0.5270159999999994


from above we can infer that areas where the percentage of households with a higher income are present the crimes are lesser . Crimes are lesser among the families with kids and both parents . Crimes are lesser among caucasian populations as well. Hence the important features make sense.

2.b.iii. How do these results compare with your results from decision trees, above?

Comparison of Cross validated performance metrics of Decision Tree and SVC
Accuracy and precision has imoproved with SVC while recall is nearly the same .
Overall SVC performs better in the data set.

3.a.REGRESSION

In [241]:
from sklearn.linear_model import LinearRegression
regression_LR=LinearRegression()
Y = clean_df['ViolentCrimesPerPop']
regression_LR.fit(X,Y)
linear_predicted=regression_LR.predict(X)
y_true = clean_df['ViolentCrimesPerPop']

3.a.ii.	What is the MSE on the training set (train on all the data then test on it all)?

In [242]:
MSE=sum((linear_predicted-Y)**2)*(1/len(Y))
print("MSE_LR = ",MSE)

MSE_LR =  0.0165167748803


3.A.i) 	Using 10-fold cross-validation, what is the estimated mean-squared-error (MSE) of the model?

In [243]:
CrossValidated_LR=cross_val_score(regression_LR, X, Y,cv=10,scoring="neg_mean_squared_error")
print ("MSE_Cross Validated: ", abs(CrossValidated_LR.mean()))

MSE_Cross Validated:  0.0200939693044


3.a.iii)	What features are most predictive of a high crime rate? A low crime rate?

In [244]:
feature_Score_reg = regression_LR.coef_
sign_features = feature_Score_reg >= 0
feature_Score_reg=abs(regression_LR.coef_)
iteration_reg= np.argsort(feature_Score_reg)[::-1]
for f in range(10):
    print("%d.  %s %f %s" % (f + 1, X.columns[iteration_reg[f]], feature_Score_reg[iteration_reg[f]], sign_features[iteration_reg[f]]))

1.  PctPersOwnOccup 0.675694 False
2.  PersPerOccupHous 0.635088 True
3.  PctHousOwnOcc 0.568133 True
4.  TotalPctDiv 0.561924 False
5.  MalePctDivorce 0.458517 True
6.  PctRecImmig8 0.432511 True
7.  MedRent 0.372728 True
8.  whitePerCap 0.351016 False
9.  PctKids2Par 0.322651 False
10.  OwnOccLowQuart 0.308170 False


Above features impact the low and high crime rates and the regression coefficients are shown as well.

Predictive features of low crime are:

1.PctPersOwnOccup: percent of people in owner occupied households

2.TotalPctDiv: percentage of population who are divorced


3.OwnOccLowQuart:owner occupied housing - lower quartile value percentage of males who are divorced

4.whitePerCap: per capita income for caucasians

5.PctKids2Par:percentage of kids in family housing with two parents

Predictive features of high crime are:

1.PctHousOwnOcc: percent of households owner occupied

2.PersPerOccupHous: mean persons per household

3.MalePctDivorce :percentage of males who are divorced


4.PctRecImmig8 : percentage of _immigrants_ who immigated within last 8 year


5.MedRent: median gross rent 

3.b.RIDGE REGRESSION
3.b.i)	What is the estimated MSE of the model under 10-fold CV

In [245]:
from sklearn.linear_model import RidgeCV
regression_RIDGE_cv = RidgeCV(alphas=(10,1,0.1,0.01,0.001),cv=10)
regression_RIDGE_cv.fit(X,Y)
ridge_cv_predicted=regression_RIDGE_cv.predict(X)
MSE=sum((ridge_cv_predicted-Y)**2)*(1/len(Y))
print("MSE_Cross Validated = ",MSE)


MSE_Cross Validated =  0.0167635291552


3.b.ii) What is the MSE on the training set (train on all the data then test on it all)

In [246]:
from sklearn.linear_model import Ridge
regression_RIDGE = Ridge(alpha= 1)
regression_RIDGE.fit(X,Y)
ridge_predicted=regression_RIDGE.predict(X)
MSE=sum((ridge_predicted-Y)**2)*(1/len(Y))
print("MSE wio CV = ",MSE)


MSE wio CV =  0.0167635291552


3.b.iii.	What is the best alpha?

In [247]:
bestAlpha=regression_RIDGE_cv.alpha_
print("Best Alpha = ",bestAlpha)

Best Alpha =  1


3.b.iv) .  What does this say about the amount of overfitting in linear regression for this problem?

The results, i.e., MSE with and without cross validation are the same  in case of ridge . NO significant change in MSE.
But cross validated MSE on linear regression is 0.019. On comparing with linear regression and ridge regression , MSE on ridge is lesser which means that overfitting is less in the model.

3.c.Quadratic Polynomial Regression

In [248]:
from sklearn.preprocessing import PolynomialFeatures
Quad_regression = PolynomialFeatures(degree=2)
quadratic = Quad_regression.fit_transform(X)
print(quadratic)

[[  1.00000000e+00   1.00000000e-02   6.10000000e-01 ...,   1.00000000e-04
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   1.00000000e-02   4.10000000e-01 ...,   1.00000000e-04
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   3.00000000e-02   3.40000000e-01 ...,   1.60000000e-03
    0.00000000e+00   0.00000000e+00]
 ..., 
 [  1.00000000e+00   3.00000000e-02   4.00000000e-01 ...,   4.00000000e-04
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   1.00000000e-02   4.50000000e-01 ...,   1.02400000e-01
    0.00000000e+00   0.00000000e+00]
 [  1.00000000e+00   1.00000000e-02   2.90000000e-01 ...,   0.00000000e+00
    0.00000000e+00   0.00000000e+00]]


3.c.i) 	What is the estimated MSE of the model under 10-fold CV?

In [249]:
quad_model=LinearRegression()
quad_cv=cross_val_score(quad_model, quadratic, Y,cv=10,scoring="neg_mean_squared_error")
print("MSE_Cross Validated = : ", abs(quad_cv.mean()))

MSE_Cross Validated = :  0.129898436576


3.c.ii.	What is the MSE on the training set (train on all the data then test on it all)

In [250]:
quad_model.fit(quadratic,Y)
quad_predicted=quad_model.predict(quadratic)
MSE=sum((quad_predicted-Y)**2)*(1/len(Y))
print("MSE wio cv = ",MSE)

MSE wio cv =  1.36976560371e-28


3.c.iii.Does this mean the quadratic model is better than the linear model for this problem?

Cross Validated MSE of linear regression is lesser than the quadratic model. So the linear model is better for this data set.

4. Dirty Data

In [251]:
df_dirty = pd.read_csv('communities-crime-full.csv')
df_dirty.head()

state county community        communityname  fold  population  \
0      8      ?         ?         Lakewoodcity     1        0.19   
1     53      ?         ?          Tukwilacity     1        0.00   
2     24      ?         ?         Aberdeentown     1        0.00   
3     34      5     81440  Willingborotownship     1        0.04   
4     42     95      6096    Bethlehemtownship     1        0.01   

   householdsize  racepctblack  racePctWhite  racePctAsian  \
0           0.33          0.02          0.90          0.12   
1           0.16          0.12          0.74          0.45   
2           0.42          0.49          0.56          0.17   
3           0.77          1.00          0.08          0.12   
4           0.55          0.02          0.95          0.09   

          ...           LandArea  PopDens  PctUsePubTrans  PolicCars  \
0         ...               0.12     0.26            0.20       0.06   
1         ...               0.02     0.12            0.45          ?   
2         ...               0.01     0.21            0.02          ?   
3         ...               0.02     0.39            0.28          ?   
4         ...               0.04     0.09            0.02          ?   

   PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0           0.04                  0.9                  0.5   
1              ?                    ?                    ?   
2              ?                    ?                    ?   
3              ?                    ?                    ?   
4              ?                    ?                    ?   

   LemasPctOfficDrugUn  PolicBudgPerPop  ViolentCrimesPerPop  
0                 0.32             0.14                 0.20  
1                 0.00                ?                 0.67  
2                 0.00                ?                 0.43  
3                 0.00                ?                 0.12  
4                 0.00                ?                 0.03  

[5 rows x 128 columns]

In [252]:
df_dirty['highCrime'] = (df_dirty['ViolentCrimesPerPop'] > 0.1).astype(int)
X = df_dirty.drop(['communityname','ViolentCrimesPerPop','highCrime'],axis=1)
X=X.apply(pd.to_numeric,errors='coerce')
X = X.fillna(X.mean()).astype(int).astype(int)
print(X)

      state  county  community  fold  population  householdsize  racepctblack  \
0         8      58      46188     1           0              0             0   
1        53      58      46188     1           0              0             0   
2        24      58      46188     1           0              0             0   
3        34       5      81440     1           0              0             1   
4        42      95       6096     1           0              0             0   
5         6      58      46188     1           0              0             0   
6        44       7      41500     1           0              0             0   
7         6      58      46188     1           0              0             0   
8        21      58      46188     1           0              0             0   
9        29      58      46188     1           0              0             0   
10        6      58      46188     1           0              0             0   
11       36      58      461

In [253]:
Y = df_dirty['highCrime']
dt_dirty = tree.DecisionTreeClassifier()
dt_dirty = dt_dirty.fit(X, Y)
dt_dirty_predicted = dt_dirty.predict(X)
pd.crosstab(df_dirty['highCrime'], dt_dirty_predicted, rownames=['highCrime'], colnames=['Predicted_dirtyData'])

Predicted_dirtyData    0     1
highCrime                     
0                    707    36
1                     40  1211

In [254]:
from sklearn.metrics import precision_score,recall_score,accuracy_score
y_true = df_dirty['highCrime']
accuracy= accuracy_score( y_true, dt_dirty_predicted )
precision = precision_score( y_true, dt_dirty_predicted )
recall = recall_score( y_true, dt_dirty_predicted )
print("Accuracy = ",accuracy,"Precision = ",precision,", Recall = ",recall)

Accuracy =  0.961885656971 Precision =  0.971130713713 , Recall =  0.968025579536


In [255]:
from sklearn.model_selection import cross_val_score
accuracy_cv = cross_val_score(dt_dirty, X, Y,cv=10,scoring="accuracy")
precision_cv= cross_val_score(dt_dirty, X, Y,cv=10,scoring="precision")
recall_cv = cross_val_score(dt_dirty, X, Y,cv=10,scoring="recall")
print ("Accuracy_CV = ", accuracy_cv.mean())
print ("Precision_CV = ", precision_cv.mean())
print ("Recall_CV = ", recall_cv.mean())

Accuracy_CV =  0.670583552089
Precision_CV =  0.736195174239
Recall_CV =  0.726685714286


4.a.	Are the CV results better or worse? What does this say about the effect of missing values?

Cross Validated Accuracy for Decision Tree Classifier in clean data set is 0.71 whereas in the dirty data set it is 0.66 and so are the precision(0.778473506115 and  0.737086806445 respectively ) and recall(0.7784 and 0.729079365079 respectively )  values
it can be seen that the values are lesser in case of dirty data set. Hence it can be concluded that the presence of missing values will have an impact on the performance metrics of any algorithms.
Though decision tree classfiers are comparitively resisitant to effect of missing values ,it can be seen that the metrics are affected.